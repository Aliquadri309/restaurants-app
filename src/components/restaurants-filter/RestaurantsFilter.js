import React, { useState } from "react";
import "./RestaurantsFilter.scss";

const RestaurantsFilter = ({ states, genres, onFilterChange }) => {
  const [stateVal, setStateVal] = useState("All");
  const [genreVal, setGenreVal] = useState("All");
  const [searchBy, setSearchBy] = useState("");

  const handleChange = (event) => {
    const { name, value } = event.target;
    if (name === "stateFilter") {
      setStateVal(value);
    } else if (name === "genreFilter") {
      setGenreVal(value);
    } else if (name === "searchBy") {
      setSearchBy(value);
    }
    if (
      name === "stateFilter" ||
      name === "genreFilter" ||
      (name === "searchBy" && (event.key === "Enter" || value === ""))
    ) {
      onFilterChange({
        stateFilter: name === "stateFilter" ? value : stateVal,
        genreFilter: name === "genreFilter" ? value : genreVal,
        searchBy: name === "searchBy" ? value : searchBy,
      });
    }
  };
  return (
    <div className="filters">
      <div className="filters__input-group">
        <label>Search By</label>
        <input
          type="text"
          className="filters__input-group__input"
          name="searchBy"
          placeholder="Name, City, Genre"
          value={searchBy}
          onKeyDown={handleChange}
          onChange={handleChange}
        />
      </div>
      <div className="filters__input-group">
        <label>Filter By State</label>
        <select
          className="filters__input-group__input"
          name="stateFilter"
          value={stateVal}
          onChange={handleChange}
        >
          {states.map((state, index) => (
            <option value={state} key={index}>
              {state}
            </option>
          ))}
        </select>
      </div>
      <div className="filters__input-group">
        <label>Filter By Genre</label>
        <select
          className="filters__input-group__input"
          name="genreFilter"
          value={genreVal}
          onChange={handleChange}
        >
          {genres.map((genre, index) => (
            <option value={genre} key={index}>
              {genre}
            </option>
          ))}
        </select>
      </div>
    </div>
  );
};

export default RestaurantsFilter;
