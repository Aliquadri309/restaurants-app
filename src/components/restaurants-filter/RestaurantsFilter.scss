.filters {
    display: flex;
    justify-content: flex-end;
    &__input-group {
        display: flex;
        flex-direction: column;
        margin-left: 1rem;
        margin-bottom: 1rem;
        label {
            font-weight: 600;
            font-size: 0.825rem;
            margin-bottom: 0.5rem;
        }
        &__input {
            padding: 0.5rem 0.75rem;
            width: 200px;
            border: 1px solid #555;
            border-radius: 0.25rem;
        }
    }
}