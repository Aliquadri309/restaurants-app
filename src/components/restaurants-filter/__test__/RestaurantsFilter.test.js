import { shallow } from "enzyme";
import toJson from "enzyme-to-json";
import RestaurantsFilter from "../RestaurantsFilter";

describe("RestaurantsFilter Component", () => {
  it("should render component without crash", () => {
    const states = ["All", "VA", "NY", "FL", "NC"];
    const genres = [
      "All",
      "American",
      "Seafood",
      "International",
      "Asian",
      "Cafe",
    ];
    const component = shallow(
      <RestaurantsFilter states={states} genres={genres} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it("should match the initial state value", () => {
    const states = ["All", "VA", "NY", "FL", "NC"];
    const genres = [
      "All",
      "American",
      "Seafood",
      "International",
      "Asian",
      "Cafe",
    ];
    const component = shallow(
      <RestaurantsFilter states={states} genres={genres} />
    );
    const searchInput = component.find(".filters__input-group__input").at(0);
    const statesInput = component.find(".filters__input-group__input").at(1);
    const genreInput = component.find(".filters__input-group__input").at(2);
    expect(searchInput.props().value).toBe("");
    expect(statesInput.props().value).toBe("All");
    expect(genreInput.props().value).toBe("All");
  });

  it("should change the state, genre, search input values on change event", () => {
    const states = ["All", "VA", "NY", "FL", "NC"];
    const genres = [
      "All",
      "American",
      "Seafood",
      "International",
      "Asian",
      "Cafe",
    ];
    const mockFun = jest.fn();
    const component = shallow(
      <RestaurantsFilter
        states={states}
        genres={genres}
        onFilterChange={mockFun}
      />
    );
    // state dropdown on change
    const statesInput = component.find(".filters__input-group__input").at(1);
    expect(statesInput.props().value).toBe("All");
    statesInput.simulate("change", {
      target: { name: "stateFilter", value: "VA" },
    });
    const statesInput1 = component.find(".filters__input-group__input").at(1);
    expect(statesInput1.props().value).toBe("VA");

    // genre dropdown on change
    const genreInput = component.find(".filters__input-group__input").at(2);
    expect(genreInput.props().value).toBe("All");
    genreInput.simulate("change", {
      target: { name: "genreFilter", value: "American" },
    });
    const genreInput1 = component.find(".filters__input-group__input").at(2);
    expect(genreInput1.props().value).toBe("American");

    // search input on change
    const searchInput = component.find(".filters__input-group__input").at(0);
    expect(searchInput.props().value).toBe("");
    searchInput.simulate("change", {
      target: { name: "searchBy", value: "city Name" },
    });
    const searchInput1 = component.find(".filters__input-group__input").at(0);
    expect(searchInput1.props().value).toBe("city Name");
  });

  it("should call filter method on input change", () => {
    const states = ["All", "VA", "NY", "FL", "NC"];
    const genres = [
      "All",
      "American",
      "Seafood",
      "International",
      "Asian",
      "Cafe",
    ];
    const mockFun = jest.fn();
    const component = shallow(
      <RestaurantsFilter
        states={states}
        genres={genres}
        onFilterChange={mockFun}
      />
    );

    const genreInput = component.find(".filters__input-group__input").at(2);
    expect(genreInput.props().value).toBe("All");
    genreInput.simulate("change", {
      target: { name: "genreFilter", value: "American" },
    });

    expect(mockFun).toHaveBeenCalled();
    expect(mockFun).toHaveBeenCalledWith({
      stateFilter: "All",
      genreFilter: "American",
      searchBy: "",
    });
  });
});
