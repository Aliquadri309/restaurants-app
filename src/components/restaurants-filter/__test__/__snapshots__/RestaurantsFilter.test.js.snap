// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RestaurantsFilter Component should render component without crash 1`] = `
<div
  className="filters"
>
  <div
    className="filters__input-group"
  >
    <label>
      Search By
    </label>
    <input
      className="filters__input-group__input"
      name="searchBy"
      onChange={[Function]}
      onKeyDown={[Function]}
      placeholder="Name, City, Genre"
      type="text"
      value=""
    />
  </div>
  <div
    className="filters__input-group"
  >
    <label>
      Filter By State
    </label>
    <select
      className="filters__input-group__input"
      name="stateFilter"
      onChange={[Function]}
      value="All"
    >
      <option
        key="0"
        value="All"
      >
        All
      </option>
      <option
        key="1"
        value="VA"
      >
        VA
      </option>
      <option
        key="2"
        value="NY"
      >
        NY
      </option>
      <option
        key="3"
        value="FL"
      >
        FL
      </option>
      <option
        key="4"
        value="NC"
      >
        NC
      </option>
    </select>
  </div>
  <div
    className="filters__input-group"
  >
    <label>
      Filter By Genre
    </label>
    <select
      className="filters__input-group__input"
      name="genreFilter"
      onChange={[Function]}
      value="All"
    >
      <option
        key="0"
        value="All"
      >
        All
      </option>
      <option
        key="1"
        value="American"
      >
        American
      </option>
      <option
        key="2"
        value="Seafood"
      >
        Seafood
      </option>
      <option
        key="3"
        value="International"
      >
        International
      </option>
      <option
        key="4"
        value="Asian"
      >
        Asian
      </option>
      <option
        key="5"
        value="Cafe"
      >
        Cafe
      </option>
    </select>
  </div>
</div>
`;
